### PR TITLE
Adapt Cluster API ClusterClass & new Tanzu Addon Manager

### PR DESCRIPTION
**Pull request by XudongLiuHarold**
Signed-off-by: Xudong Liu <xudongl@vmware.com>

**What this PR does / why we need it**:

- Make AKO Operator can work together with cluster API cluster class
- Make AKO Operator can work together with new tanzu addon manager

**Which issue(s) this PR fixes**:

Fixes #

**Describe testing done for PR**:

TODO

**Special notes for your reviewer**:

**Release note**:

```release-note

```
**New PR Checklist**

- [x] Ensure PR contains only public links or terms
- [x] Use good commit [messages](https://github.com/vmware-tanzu/tanzu-framework/blob/main/CONTRIBUTING.md)
- [x] Squash the commits in this branch before merge to preserve our git history
- [ ] If this PR is just an idea or POC, use a [Draft PR](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/about-pull-requests#draft-pull-requests) instead of a full PR
- [x] Add appropriate [labels](https://github.com/vmware-tanzu/load-balancer-operator-for-kubernetes/labels) according to what type of issue is being addressed.

### CODE REVIEW

File: controllers/akodeploymentconfig/cluster/cluster_controller_addon_secret.go [hunk 1]
```diff
@@ -63,7 +65,18 @@ func (r *ClusterReconciler) ReconcileAddonSecret(
...
 	}
 	secret = newAddonSecret.DeepCopy()
-	return res, r.Update(ctx, secret)
+	if err := r.Update(ctx, secret); err != nil {
+		log.Error(err, "Failed to update ako add on secret, requeue")
+		return res, err
+	}
+
+	// patch cluster bootstrap here
+	if err := r.patchAkoPackageRefToClusterBootstrap(ctx, cluster); err != nil {
```
Comment (HanFa) [inline controllers/akodeploymentconfig/cluster/cluster_controller_addon_secret.go:74]: i know its from previous design discussion, but why can't we deploy a TKR where the ClusterBoostrapTemplate has ako as additionalpackage?

File: pkg/ako-operator/lib_test.go [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+	"os"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	v1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+)
+
+var _ = Describe("AKO Operator lib unit test", func() {
```
Comment (HanFa) [inline pkg/ako-operator/lib_test.go:17]: kudos for providing these tests!

File: pkg/ako-operator/lib.go [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+const (
+	// legacy cluster environment variables
+	// DeployInBootstrapCluster - defines if ako operator is deployed in bootstrap cluster
+	DeployInBootstrapCluster = "bootstrap_cluster"
+
+	// IsControlPlaneHAProvider - defines if ako operator is going to provide control plane HA
+	IsControlPlaneHAProvider = "avi_control_plane_ha_provider"
+
+	// ManagementClusterName - defines the management cluster name ako operator running in
+	ManagementClusterName = "tkg_management_cluster_name"
```
Comment (HanFa) [inline pkg/ako-operator/lib.go:24]: unused variable?

File: pkg/ako-operator/lib.go [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+	// ControlPlaneEndpointPort - defines the control plane endpoint port
+	ControlPlaneEndpointPort = "control_plane_endpoint_port"
+
+	// cluster class cluster environment variables
+	// KubeVipLoadBalancerProvider - defines if cluster using kube-vip to implement load balancer
+	// type of service
+	KubeVipLoadBalancerProvider = "kubeVipLoadBalancerProvider"
```
Comment (HanFa) [inline pkg/ako-operator/lib.go:32]: renamed to `kube_vip_load_balancer_provider` for consistency

File: pkg/ako-operator/lib.go [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+				err := json.Unmarshal(clusterVariable.Value.Raw, &apiServerPort)
+				if err != nil || !validatePortNumber(apiServerPort) {
+					return 6443
+				}
+			}
+		}
+		return int32(apiServerPort)
+	} else {
+		apiServerPort, err := strconv.Atoi(os.Getenv(ControlPlaneEndpointPort))
+		println("apiServerPort")
```
Comment (HanFa) [inline pkg/ako-operator/lib.go:101]: remove the debug dump, maybe add some logs even better

File: controllers/configmap/configmap_controller.go [hunk 6]
```diff
@@ -56,6 +56,18 @@ func (r *ConfigMapReconciler) initAVI(ctx context.Context,
 			log.Error(err, "Cannot init AVI clients from secrets")
 			return res, err
 		}
+		version, err := r.aviClient.GetControllerVersion()
```
Comment (lubronzhan) [inline controllers/configmap/configmap_controller.go:59]: This is the bug fix right, reinitialize the client

File: pkg/ako-operator/lib.go [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+}
+
+func IsLoadBalancerProvider(cluster *clusterv1.Cluster) bool {
+	if IsClusterClassBasedCluster(cluster) {
+		for _, clusterVariable := range cluster.Spec.Topology.Variables {
+			if clusterVariable.Name == KubeVipLoadBalancerProvider {
+				var kubeVipLoadBalancerProvider bool
+				if err := json.Unmarshal(clusterVariable.Value.Raw, &kubeVipLoadBalancerProvider); err == nil {
+					return !kubeVipLoadBalancerProvider
+				}
```
Comment (lubronzhan) [inline pkg/ako-operator/lib.go:76]: Log the error as well?

File: pkg/ako-operator/lib.go [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+		}
+	}
+	return true
+}
+
+func validatePortNumber(number int) bool {
+	return number > 0 && number < 65536
+}
+
+func GetControlPlaneEndpointPort(cluster *clusterv1.Cluster) int32 {
```
Comment (lubronzhan) [inline pkg/ako-operator/lib.go:87]: apiServerEndpoint is also defined in Cluster variable now.
Should we also extract endpoint and put it inside contorl plane HA service?

File: pkg/ako-operator/lib.go (hunk 3 above)
Comment (XudongLiuHarold) [reply pkg/ako-operator/lib.go:24]: Yep, removing it

File: controllers/akodeploymentconfig/cluster/cluster_controller_addon_secret.go [hunk 9]
```diff
@@ -159,3 +185,52 @@ func (r *ClusterReconciler) getClusterAviUserSecret(cluster *clusterv1.Cluster,
...
+func (r *ClusterReconciler) patchAkoPackageRefToClusterBootstrap(ctx context.Context, cluster *clusterv1.Cluster) error {
+	bootstrap, err := r.getClusterBootstrap(ctx, cluster)
+	if err != nil {
+		return err
+	}
+
+	akoClusterBootstrapPackage := &runv1alpha3.ClusterBootstrapPackage{
+		RefName: akoov1alpha1.AkoClusterBootstrapRefName,
+		ValuesFrom: &runv1alpha3.ValuesFrom{
+			SecretRef: r.akoAddonSecretName(cluster),
```
[the remark this answers is not in the document]
Comment (XudongLiuHarold) [reply controllers/akodeploymentconfig/cluster/cluster_controller_addon_secret.go:212]: It is 253 characters

File: controllers/configmap/configmap_controller.go (hunk 6 above)
Comment (XudongLiuHarold) [reply controllers/configmap/configmap_controller.go:59]: Yep

File: pkg/ako-operator/lib.go (hunk 8 above)
Comment (XudongLiuHarold) [reply pkg/ako-operator/lib.go:87]: I see, we should

**Review by chenlin07 — APPROVED** [review]

**Pull request merged**